client/watch.go

```diff
@@ -0,0 +1,98 @@
+// Copyright 2023 CloudWeGo Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"fmt"
+
+	"github.com/cloudwego/kitex/pkg/klog"
+	"github.com/kitex-contrib/config-file/parser"
+	"github.com/kitex-contrib/config-file/utils"
+)
+
+type ConfigWatcher struct {
+	config      *parser.ClientFileConfig
+	callbacks   []func()
+	key         string
+	from        string // client name
+	to          string // server name
+	filewatcher *utils.FileWatcher
+}
+
+// NewConfigWatcher init a watcher for the config file
+func NewConfigWatcher(filepath, from, to string) *ConfigWatcher {
+	fw, err := utils.NewFileWatcher(filepath)
+	if err != nil {
+		panic(err)
+	}
+
+	return &ConfigWatcher{
+		filewatcher: fw,
+		key:         fmt.Sprintf("%s%s%s", from, "/", to),
```

Comment:
This implementation limits the granularity of the config. For example, in ByteDance, a service are likely to be deployed in several clusters, and there's need to set different timeout/retry/circuitbreaker config for each target cluster.

Reply:
So we should define an interface for the config key, for example:

```go
type KeyProvider interface {
    GetKey(RPCInfo) string
    ParseKey(key string) Key
}

type Key struct {
    ClientName string
    ServerName string
    Tags map[string]string
}
```

And our default implementation could be as the example ("$client/$server")

Reply:
We may assume (and informs users in the doc) that each client should be responsible for only one `Key`, to simplify our implementation. This way we can unify the clientWatcher and serviceWatcher, with different keys.